### PR TITLE
Add libxml2 as a dedicated report formatter

### PR DESCRIPTION
**Pull request by ar-cetitec**
Selected by -X/--libxml for cgreen-runner, the report formatter uses libxml2 to construct an XML with report data. The libxml2 provides a complete and well-supported formatting of XML and UTF-8, and is better compatible with existing analyzing software, at the expense of build and run-time dependency on the libxml2 library.

### CODE REVIEW

**Comment by thoni56** [comment]
This is great work! Thanks.

I think it is the right call to use libxml, but I'm pondering (pros/cons for...]

- should be just replace the old xml_printer with this?
- could we automagically select to use libxml_printer if libxml is available, or even not include the xml_printer if it's not?

I'm thinking that if you want XML-output you are probably in a "bigger"/server/CI environment and would have no problem with the extra dependency. But if you are in an embedded(-like) environment you probably don't want that extra dependency, but, again probably, don't need XML output either.

Thoughts?

**Comment by ar-cetitec** [comment]
> I think it is the right call to use libxml, but I'm pondering (pros/cons for...)
> 
> * should we just replace the old xml_printer with this?

Personally, I'd prefer to have both: their output differs. People might have the samples of the output of the non-libxml2 reporter and use them to compare with the output of cgreen-runner as plain text. And it might differ even after re-formatting into canonical XML. This change has a potential to break someones testing automation.

> * could we automagically select to use libxml_printer if libxml is available, or even not include the xml_printer at all if it's not?

It should be possible to enable it automagically: the code just has to verify presence of `LibXml2_FOUND` after `find_package`.
 
> I'm thinking that if you want XML-output you are probably in a "bigger"/server/CI environment and would have no problem with the extra dependency.

I don't see an additional dependency as an argument of a big or small environment. It is the administrative effort of maintaining it which makes new dependencies worth an extra minute of consideration.

> But if you are in an embedded(-like) environment you probably don't want that extra dependency, but, again probably, don't need XML output either.

In the latter case, an option to disable that reporter might be useful.

**Comment by thoni56** [comment]
Good comments!

> > I think it is the right call to use libxml, but I'm pondering (pros/cons for...)
> > 
> > * should we just replace the old xml_printer with this?
> 
> Personally, I'd prefer to have both: their output differs. People might have the samples of the output of the non-libxml2 reporter and use them to compare with the output of cgreen-runner as plain text. And it might differ even after re-formatting into canonical XML. This change has a potential to break someones testing automation.

Right. So we might consider removing the old xml_printer but no earlier than 2.0.

> > * could we automagically select to use libxml_printer if libxml is available, or even not include the xml_printer at all if it's not?
> 
> It should be possible to enable it automagically: the code just has to verify presence of `LibXml2_FOUND` after `find_package`.

I think we should do this. It lessens the "administrative burdon". Could you please add that?

> > But if you are in an embedded(-like) environment you probably don't want that extra dependency, but, again probably, don't need XML output either.
> 
> In the latter case, an option to disable that reporter might be useful.

This is a good option. Then in "normal" circumstances, where you possibly already have libxml2 installed you get libxml_printer automatically and if you are in an environment that doesn't have it you don't. But in case you are somehow running the tests in an environment that do have it but you, for some reason, memory, library restrictions, ... can't or don't want it you can disable building it.

I think with those two additions this could could be merged.

Again, good job!

**Comment by ar-cetitec** [comment]
> > > * could we automagically select to use libxml_printer if libxml is available, or even not include the xml_printer at all if it's not?
> > It should be possible to enable it automagically: the code just has to verify presence of `LibXml2_FOUND` after `find_package`.
> I think we should do this. It lessens the "administrative burdon". Could you please add that?

Done.

**Comment by ar-cetitec** [comment]
> > > But if you are in an embedded(-like) environment you probably don't want that extra dependency, but, again probably, don't need XML output either.
> > 
> > In the latter case, an option to disable that reporter might be useful.
> 
> This is a good option. Then in "normal" circumstances, where you possibly already have libxml2 installed you get libxml_printer automatically and if you are in an environment that doesn't have it you don't. But in case you are somehow running the tests in an environment that do have it but you, for some reason, memory, library restrictions, ... can't or don't want it you can disable building it.
> 

Now this one turned out quite extensive. A second pair of eyes might be useful.

**Comment by ar-cetitec** [comment]
It might make sense to add testing without xml/libxml reporters to Travis. What do you think?

**Comment by thoni56** [comment]
> It might make sense to add testing without xml/libxml reporters to Travis. What do you think?

Since we now have build logic around this, I agree. Do we need that to be an extra matrix entry? That would double the build times... Any other ideas?

File: tools/cgreen-runner.c [hunk 1]
```diff
@@ -32,16 +32,25 @@ static void version(void) {
 /*----------------------------------------------------------------------*/
 static void usage(const char *program_name) {
     printf("cgreen-runner for Cgreen unittest and mocking framework v%s\n\n", VERSION);
-    printf("Usage:\n    %s [--xml <prefix>] [--suite <name>] [--verbose] [--quiet] [--no-run] [--help] (<library> [<test>])+\n\n", program_name);
+    printf("Usage:\n    %s [--suite <name>] [--verbose] [--quiet] [--no-run] "
+#if HAVE_XML_REPORTER
+           "--xml <prefix>"
```
Comment (thoni56) [inline tools/cgreen-runner.c:37]: You need a space at the end here, as well as in the --libxml2 line.

Comment (ar-cetitec) [reply tools/cgreen-runner.c]: Fixed.

**Comment by ar-cetitec** [comment]
> > It might make sense to add testing without xml/libxml reporters to Travis. What do you think?
> 
> Since we now have build logic around this, I agree. Do we need that to be an extra matrix entry? That would double the build times... Any other ideas?

That would likely more than double the build times. Maybe build and test less-featured configurations first in the `script:` section, using incremental re-configuration and build for the final full-featured (and the default) build?

Something like this (assuming script jobs fail if a command fails):
```diff --git a/.travis.yml b/.travis.yml
index 917c647..c124477 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -30,6 +30,8 @@ before_install:
 script:
   - mkdir -p build
   - cd build
+  - cmake -DCGREEN_WITH_XML:BOOL=OFF -DCGREEN_WITH_LIBXML2:BOOL=OFF -DCGREEN_WITH_UNIT_TESTS:BOOL=ON -DCGREEN_WITH_STATIC_LIBRARY:BOOL=$CGREEN_WITH_STATIC_LIBRARY -DCGREEN_INTERNAL_WITH_GCOV:BOOL=OFF ..
+  - make -j2
   - cmake -DCGREEN_WITH_UNIT_TESTS:BOOL=ON -DCGREEN_WITH_STATIC_LIBRARY:BOOL=$CGREEN_WITH_STATIC_LIBRARY -DCGREEN_INTERNAL_WITH_GCOV:BOOL=$WITH_GCOV ..
   - make -j2
   - ctest --output-on-failure
```

**Comment by thoni56** [comment]
> Something like this (assuming script jobs fail if a command fails):

Yes, that sounds like a better idea. Have you tried it? If it works, go for it.

**Comment by ar-cetitec** [comment]
> > Something like this (assuming script jobs fail if a command fails):

Turned out that it is [not quite like this](https://docs.travis-ci.com/user/job-lifecycle/#customizing-the-build-phase): Travis will continue execution of the jobs commands, but the exit status will be recorded and the job will be report as failed.



Which means that the current job code runs "make" and "ctest" even if "cmake" command failed. Given that the CI time is free (at present, at least), the only problem with that is a cluttered build log.

Considering that problem annoying enough, I rearranged all commands to only run tests if the code is successfully built.

> 
> Yes, that sounds like a better idea. Have you tried it? If it works, go for it.
>

Done.

**Comment by ar-cetitec** [comment]
Hi! The last comment in this PR has been a month ago. Is something wrong with the PR?

**Comment by thoni56** [comment]
> Hi! The last comment in this PR has been a month ago. Is something wrong with the PR?

Just very busy with other engagements. Sorry.

**Comment by ar-cetitec** [comment]
Thanks!

**Pull request merged**